compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+            ntimes(times) { run =>
+              val start = System.nanoTime()
+              val rep = super.doCompile(comp, files)
+              ctx.echo(s"\ntime run $run: ${(System.nanoTime - start) / 1000000}ms")
+              rep
+            }
+        }
+
+      val allArgs = addOutDir(flags)
+      driver.process(allArgs ++ files.map(_.getAbsolutePath), reporter = reporter)
+
+      // If the java files failed compilation before, we try again after:
+      if (!javaCompiledBefore)
+        assert(compileWithJavac(javaFiles), s"java compilation failed for ${javaFiles.mkString(", ")}")
+
+      reporter
+    }
+
+    private[ParallelTesting] def execute(): this.type = {
+      assert(_targetsCompiled == 0, "not allowed to re-use a `CompileRun`")
+
+      if (allTargets.nonEmpty) {
+        val pool = threadLimit match {
+          case Some(i) => JExecutors.newWorkStealingPool(i)
+          case None => JExecutors.newWorkStealingPool()
+        }
+
+        if (interactive && !suppressAllOutput) pool.submit(statusRunner)
+
+        allTargets.foreach { target =>
+          pool.submit(compilationRunnable(target))
+        }
+
+        pool.shutdown()
+        if (!pool.awaitTermination(10, TimeUnit.MINUTES))
+          throw new TimeoutException("Compiling targets timed out")
+
+        if (didFail) {
+          echo {
+            """|
+               |================================================================================
+               |Test Report
+               |================================================================================
+               |Failing tests:""".stripMargin
+          }
+          failedCompilationTargets.toArray.sorted.foreach(echo)
+          failureInstructions.iterator.foreach(echo)
+        }
+      }
+      else echo {
+        regex
+          .map(r => s"""No files matched regex "$r" in test""")
+          .getOrElse("No tests available under target - erroneous test?")
+      }
+
+      this
+    }
+  }
+
+  @inline private final def isCompilable(f: JFile): Boolean = {
+    val name = f.getName
+    name.endsWith(".scala") || name.endsWith(".java")
+  }
+
+  private final class PosTest(targets: List[Target], times: Int, threadLimit: Option[Int], suppressAllOutput: Boolean)
+  extends Test(targets, times, threadLimit, suppressAllOutput) {
+    protected def compilationRunnable(target: Target): Runnable = new Runnable {
+      def run(): Unit = compileTry {
+        target match {
+          case ConcurrentCompilationTarget(files, flags, outDir) => {
+            val sourceFiles = files.filter(isCompilable)
+            val reporter = compile(sourceFiles, flags, false, outDir)
+            completeCompilation(reporter.errorCount)
+
+            if (reporter.errorCount > 0) {
+              fail()
+              val errorMsg = target.buildInstructions(reporter.errorCount, reporter.warningCount)
```

Comment:
Those lines are duplicated 6 times in various places. Extract to a helper method?